firedrake/mesh.py

```diff
@@ -1537,6 +1691,157 @@ def ExtrudedMesh(mesh, layers, layer_height=None, extrusion_type='uniform', kern
     return self
 
 
+def VertexOnlyMesh(mesh, vertexcoords, comm=COMM_WORLD):
+    """
+    Create a vertex only mesh, immersed in a given mesh, with vertices
+    defined by a list of coordinates.
+
+    :arg mesh: The unstructured mesh in which to immerse the vertex only
+    mesh.
+    :arg vertexcoords: A list of coordinate tuples which defines the vertices.
+    :kwarg comm: Optional communicator to build the mesh on (defaults to
+        COMM_WORLD).
+    """
+
+    import firedrake.functionspace as functionspace
+    import firedrake.function as function
+
+    mesh.init()
+
+    vertexcoords = np.asarray(vertexcoords, dtype=np.double)
+    gdim = mesh.geometric_dimension()
+    tdim = mesh.topological_dimension()
+    pdim = np.shape(vertexcoords)[1]
+    if pdim != tdim:
+        raise ValueError(f"Mesh topological dimension {tdim} must match point list dimension {pdim}")
+
+    if isinstance(mesh.topology, ExtrudedMeshTopology):
+        raise NotImplementedError("Extruded meshes are not supported")
+
+    if mesh.coordinates.function_space().ufl_element().degree() > 1:
+        raise NotImplementedError("Only straight edged meshes are supported")
+
+    swarm = _pic_swarm_in_plex(mesh.topology._plex, vertexcoords, fieldnames=["parentcellnum"], blocksizes=[1])
+
+    dmswarm.label_pic_parent_cell_nums(swarm, mesh)
+
+    # Topology
+    topology = VertexOnlyMeshTopology(swarm, mesh.topology, name="swarmmesh", reorder=False)
+
+    # Geometry
+    tcell = topology.ufl_cell()
+    cell = tcell.reconstruct(geometric_dimension=gdim)
+    element = ufl.VectorElement("DG", cell, 0)
+    # Create mesh object
+    vmesh = MeshGeometry.__new__(MeshGeometry, element)
+    vmesh._topology = topology
+    vmesh._parent_mesh = mesh
+
+    def callback(self):
+        """Finish initialisation."""
+        del self._callback
+        # Finish the initialisation of mesh topology
+        self.topology.init()
+
+        coordinates_fs = functionspace.VectorFunctionSpace(self.topology, "DG", 0,
+                                                           dim=gdim)
+
+        coordinates_data = dmswarm.reordered_coords(swarm, coordinates_fs.dm.getDefaultSection(),
+                                                    (self.num_vertices(), gdim))
+
+        coordinates = function.CoordinatelessFunction(coordinates_fs,
+                                                      val=coordinates_data,
+                                                      name="Coordinates")
+
+        self.__init__(coordinates)
```

Comment:
Much of this is the same as `Mesh(...)` with the difference that you access stuff from `dmswarm` not `dmplex`. Can we refactor to share code?

I worry that this adds a second place that one has to check when modifying anything to do with meshes.

Reply:
So the common stuff is at the end of Mesh: 
```python
def Mesh(...)
   ...

    # Create mesh topology
    topology = MeshTopology(plex, name=name, reorder=reorder,
                            distribution_parameters=distribution_parameters)

    tcell = topology.ufl_cell()
    if geometric_dim is None:
        geometric_dim = tcell.topological_dimension()
    cell = tcell.reconstruct(geometric_dimension=geometric_dim)

    element = ufl.VectorElement("Lagrange", cell, 1)
    # Create mesh object
    mesh = MeshGeometry.__new__(MeshGeometry, element)
    mesh._topology = topology

    def callback(self):
        """Finish initialisation."""
        del self._callback
        # Finish the initialisation of mesh topology
        self.topology.init()

        coordinates_fs = functionspace.VectorFunctionSpace(self.topology, "Lagrange", 1,
                                                           dim=geometric_dim)

        coordinates_data = dmplex.reordered_coords(plex, coordinates_fs.dm.getDefaultSection(),
                                                   (self.num_vertices(), geometric_dim))

        coordinates = function.CoordinatelessFunction(coordinates_fs,
                                                      val=coordinates_data,
                                                      name="Coordinates")

        self.__init__(coordinates)

    mesh._callback = callback
    return mesh
```
I suppose I could create some functions to do the common stuff with settable function space and and degree? Something like
```python
def _make_geometry(...)
    ...

def _make_topology(...)
    ...
def _make_callback(...)
    # Generate callback for Mesh and VertexOnlyMesh
    ...
```
so you'd have one fewer place to modify. `ExtrudedMesh` is done differently though (and is already a second place you have to update stuff) and I'd rather not fiddle with the implementation of extrusion in this PR.

Reply:
I realise again I don't think this two-phase initialisation routine is really necessary for these meshes, but meh.

Reply:
2 stage initialisation removed so resolved.